### ISSUE TITLE
(PARKED) New Rule: border-contrast-component

### ISSUE THREAD

**Issue by jkodu** [dequelabs/axe-core#854]
# border-contrast-component

The rule should test the contrast of `input`, `select` & `textarea` elements.

**Tags:** wcag21aa, wcag1411

## Matcher

`input`, `select` & `textarea` elements. 

## Checks

Check 1: (contrast-border-on-outside-background)
- Contrast between container background-color and form element border is at-least 3:1.

Check 2: (contrast-border-on-inside-background)
- Contrast between from element border and background-color is at-least 3:1.

Check 3: (contrast-border-background-on-background)
- Contrast between background-color of container and form element's background-color is at-least 3:1.

All 3 checks above are under the same rule and they are `any` checks.

## How to Implement
- Do the above checks and report as necessary, otherwise raise as a needs review.

Conditions for which to raise `Needs Review`:
- Could not determine a border.
- Could not determine a background.
- Input element with color set to alpha 0.
- Opacity to 0.

## Background
- https://www.w3.org/TR/WCAG21/#non-text-contrast

### Passed

```html
<input type="text" style='border: 0; background: #666; color: #fff;'>
```

### Failed

```html
<input type="text" style="border: 1px solid #eee; background-color: #efefef; color: #000">
```

### Inapplicable

```html
<div contenteditable=true role='textbox'>
<input type="text" disabled>
<input type="text" aria-hidden='true'>
<input type="hidden">
<input type="checkbox">
<input type="radio">
<input type="button">
<input type="image">
<input type="submit">
<input type="reset">
```

**Comment by WilcoFiers** [comment]
Check should be that the border and/or background has sufficient color contrast with the background around the text element.

**Comment by jkodu** [comment]
work for this has commenced in branch - https://github.com/dequelabs/axe-core/tree/rule-contrast-border-input-select-textarea

**Comment by dylanb** [comment]
This case should not be inapplicable because it will still be visible to (be able to be interacted with by sighted users).

`<input type="text" aria-hidden='true'>`

**Comment by dylanb** [comment]
Can users change the display of `select` elements? I don't think its possible, so `select` elements should be excluded

What about `<summary>`? Is it possible to style them?

**Comment by jkodu** [comment]
@WilcoFiers 
To confirm edge cases around inputs styled like links/ summary

**Comment by WilcoFiers** [comment]
OK  so here's the thing. I think this rule can be broader than form fields. In fact, this rule is pretty much applicable to any widget that has its border style changed to something that isn't the browser default. If there is no border, such as for the summary element, it should not be applicable. If there is a border it should be tested.

**Comment by jkodu** [comment]
As WCAG description has changed for this SC, closing this.
